support/kwok/kwok.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kwok
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	log "k8s.io/klog/v2"
+
+	"github.com/vladimirvivien/gexe"
+)
+
+var kwokVersion = "v0.2.0"
+
+type Cluster struct {
+	name        string
+	e           *gexe.Echo
+	kubecfgFile string
+	version     string
+}
+
+func NewCluster(name string) *Cluster {
+	return &Cluster{name: name, e: gexe.New()}
+}
+
+// WithVersion set kind version
+func (k *Cluster) WithVersion(ver string) *Cluster {
+	k.version = ver
+	return k
+}
+
+func (k *Cluster) getKubeconfig() (string, error) {
+	kubecfg := fmt.Sprintf("%s-kubecfg", k.name)
+
+	p := k.e.RunProc(fmt.Sprintf(`kwokctl get kubeconfig --name %s`, k.name))
+	if p.Err() != nil {
+		return "", fmt.Errorf("kwokctl get kubeconfig: %w", p.Err())
+	}
+
+	var stdout bytes.Buffer
+	if _, err := stdout.ReadFrom(p.Out()); err != nil {
+		return "", fmt.Errorf("kwokctl kubeconfig stdout bytes: %w", err)
+	}
+
+	file, err := os.CreateTemp("", fmt.Sprintf("kwok-cluster-%s", kubecfg))
+	if err != nil {
+		return "", fmt.Errorf("kwok kubeconfig file: %w", err)
+	}
+	defer file.Close()
+
+	k.kubecfgFile = file.Name()
+
+	if n, err := io.Copy(file, &stdout); n == 0 || err != nil {
+		return "", fmt.Errorf("kwok kubecfg file: bytes copied: %d: %w]", n, err)
+	}
+
+	return file.Name(), nil
+}
+
+func (k *Cluster) clusterExists(name string) (string, bool) {
+	clusters := k.e.Run("kwokctl get clusters")
+	for _, c := range strings.Split(clusters, "\n") {
+		if c == name {
+			return clusters, true
+		}
+	}
+	return clusters, false
+}
+
+func (k *Cluster) CreateWithConfig(kwokConfigFile string) (string, error) {
+	return k.Create("--config", kwokConfigFile)
+}
+
+func (k *Cluster) Create(args ...string) (string, error) {
+	log.V(4).Info("Creating kwok cluster ", k.name)
+	if err := k.findOrInstallKwok(k.e); err != nil {
+		return "", err
+	}
+
+	if _, ok := k.clusterExists(k.name); ok {
+		log.V(4).Info("Skipping kwok Cluster.Create: cluster already created: ", k.name)
+		return k.getKubeconfig()
+	}
+
+	command := fmt.Sprintf(`kwokctl create cluster --name %s`, k.name)
```

Comment:
by default, this would spin up a k8s cluster using docker-compose, if we want to default it to kind, need to pass-in `--runtime kind` explicitly.

Reply:
@Huang-Wei does `--runtime kind` would also require `kind` to be installed? thinking if we need to add kind dependency as well when installing kwok

Reply:
@wzshiming `kwokctl` wouldn't install the missing runtime binary, right?

Reply:
@Huang-Wei @wzshiming do we want to default it to `kind`? or we can let users pass CreateWithConfig(kwokConfigFile string) where file has entry for `runtime: kind`

Reply:
I'm fine with defaulting the same default runtime (docker) in kwokctl here; unless e2e-framework is keen on defaulting to kind.

Reply:
My suggestion: since e2e-framework already has support for kind clusters, let's default to kind. If someone wants to use docker-compose, they can do so using the kwok config file.

Reply:
Hi @vladimirvivien since `kwokctl` defaults to using `docker-compose` I think it makes more sense to default to `docker-compose`. if we default to `kind` with `--runtime kind` argument we will not be able to override to docker using the kwok config file since `--runtime kind` takes precendence over kwok config file